specs/sharding/beacon-chain.md

```diff
@@ -810,6 +812,38 @@ def process_shard_proposer_slashing(state: BeaconState, proposer_slashing: Shard
     slash_validator(state, proposer_index)
 ```
 
+#### `process_shard_basefees`
+
+```python
+def process_shard_basefee_updates(state: BeaconState) -> None:
+    if state.slot < SLOTS_PER_BASE_FEE_UPDATES:  # no base fee processing on initial stage
+        return
+    
+    # The last SLOTS_PER_BASE_FEE_UPDATES slots are considered in the adjustment.
+    # Since inclusion can be late and out of order, and missing beacon blocks should not affect shard capacity as much.
+
+    start_slot = state.slot - SLOTS_PER_BASE_FEE_UPDATES
+    for slot in range(start_slot, state.slot):
+      buffer_index = slot % SHARD_STATE_MEMORY_SLOTS
+      shard_count = len(state.shard_buffer[buffer_index])
+      for shard_index in range(shard_count):
+          committee_work = state.shard_buffer[buffer_index][shard_index]
+  
+          samples_count = 0
+          if committee_work.status.selector == SHARD_WORK_CONFIRMED:
+              attested: AttestedDataCommitment = committee_work.status.value
+              samples_count = attested.commitment.samples_count
+          elif committee_work.status.selector == SHARD_WORK_PENDING:
+              pending: List[PendingShardHeader, MAX_SHARD_HEADERS_PER_SHARD] = committee_work.status.value
+              # Conflicting headers add up to relieve network stress.
+              samples_count = sum(header.attested.commitment.samples_count for header in pending)
```

Comment:
Also not sure if the price should be adjusted for `PENDING` headers at all.
What if just make the price adjustment once per blob right upon its confirmation? From my perspective that could look simpler and more straightforward

Reply:
If a proposer makes multiple proposals, then there is capacity of 4 of them (this pending-headers list is specific to a single shard+slot combination), to try keep attestations useful, even if the proposer messes up / attacks.  But 4 proposals also means more network stress, so we sum up the worst-case claimed capacity.

Reply:
Also: base fee, capacity, sharding performance and confirmations are very related. If the system is unhealthy, I prefer to smooth into adjustments based on the latest view, instead of waiting for confirmation.

Reply:
I mean that function `compute_updated_sample_price()` expects `samples_count <= MAX_SAMPLES_PER_BLOB` else it would not satisfy adjustment requirement: 
```
SAMPLE_PRICE_ADJUSTMENT_COEFFICIENT | Sample price may decrease/increase by at most exp(1 / this value) per epoch
```

Reply:
BTW, may be it makes sense to discuss again the necessity of handling `>1` headers from the same proposer? 

Producing `>1` headers is definitely either a malicious of malfunctioning behavior. It supposed to be a pretty exceptional event, because of slashing. I still couldn't get how accounting of such cases may help with network liveness.

Restricting sate to have just one pending header per slot+shard would simplify many things IMO